internal/encryption/pkcs11.go

```diff
@@ -0,0 +1,194 @@
+	result := make([]byte, len(iv)+len(ciphertext))
+	copy(result[:ivSize], iv)
+	copy(result[ivSize:], ciphertext)
+
+	return result, nil
+}
+
+// Decrypt decrypts the ciphertext using AES-CBC algorithm.
+func (h *PKCS11Backend) Decrypt(ciphertext []byte) ([]byte, error) {
+	if len(ciphertext) < ivSize {
+		return nil, fmt.Errorf("invalid ciphertext: too short to contain IV")
+	}
+
+	// Extract IV from the start of the combined data
+	iv := ciphertext[:ivSize]
+	ciphertext = ciphertext[ivSize:]
+
+	session, key, err := h.connectToBackend()
+	if err != nil {
+		return nil, fmt.Errorf("failed to connect to backend: %w", err)
+	}
+	defer func() {
+		if cleanupErr := h.cleanupSession(session); cleanupErr != nil {
+			fmt.Printf("Warning: cleanup errors during decryption: %v\n", cleanupErr)
+		}
+	}()
+
+	mech := pkcs11.NewMechanism(pkcs11.CKM_AES_CBC_PAD, iv)
+
+	if err := h.ctx.DecryptInit(session, []*pkcs11.Mechanism{mech}, key); err != nil {
+		return nil, fmt.Errorf("failed to initialize decryption: %w", err)
+	}
+
+	plaintext, err := h.ctx.Decrypt(session, ciphertext)
+	if err != nil {
+		return nil, fmt.Errorf("failed to decrypt data: %w", err)
+	}
+	return plaintext, nil
+}
+
+func (h *PKCS11Backend) findKey(session pkcs11.SessionHandle, id uint16) (pkcs11.ObjectHandle, error) {
+	keyIDBytes := []byte{byte(id >> 8), byte(id & 0xff)}
+	template := []*pkcs11.Attribute{
+		pkcs11.NewAttribute(pkcs11.CKA_CLASS, pkcs11.CKO_SECRET_KEY),
+		pkcs11.NewAttribute(pkcs11.CKA_ID, keyIDBytes),
+	}
+
+	if err := h.ctx.FindObjectsInit(session, template); err != nil {
+		return 0, fmt.Errorf("failed to initialize object search: %w", err)
+	}
+	objects, _, err := h.ctx.FindObjects(session, 1)
+	if err != nil || len(objects) == 0 {
+		return 0, fmt.Errorf("failed to find objects: %w", err)
+	}
+	if err := h.ctx.FindObjectsFinal(session); err != nil {
+		return 0, fmt.Errorf("failed to finalize object search: %w", err)
+	}
+	return objects[0], nil
+}
+
+// cleanupSession performs cleanup of an active session
+func (h *PKCS11Backend) cleanupSession(session pkcs11.SessionHandle) error {
+	if err := h.ctx.Logout(session); err != nil {
+		return err
+	}
+	if err := h.ctx.CloseSession(session); err != nil {
+		return err
+	}
+	if err := h.ctx.Finalize(); err != nil {
+		return err
+	}
+	return nil
+}
+
+// cleanupInitialization performs cleanup of just the initialization
+func (h *PKCS11Backend) cleanupInitialization() error {
+	if err := h.ctx.Finalize(); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (h *PKCS11Backend) connectToBackend() (pkcs11.SessionHandle, pkcs11.ObjectHandle, error) {
+	if err := h.ctx.Initialize(); err != nil {
+		return 0, 0, err
+	}
+
+	slots, err := h.ctx.GetSlotList(true)
+	if err != nil || len(slots) == 0 {
+		if cleanupErr := h.cleanupInitialization(); cleanupErr != nil {
+			fmt.Printf("Warning: cleanup error after GetSlotList failure: %v\n", cleanupErr)
+		}
+		return 0, 0, err
+	}
+	slot := slots[0]
```

Comment:
I don't know all that much about hsm's yet but is there good reason to assume the first slot?

Reply:
Good point. This would have almost always worked for yubihsm2, but it nothing guarantees it always works for any hsm. hsms can have multiple slots, each slot will have a different token and our object (encryption key) might not be in the first slot. Since the order of the slots is not always necessarily the same, and slots are not configured by user (as far as I understood) I took an approach of trying to find the key in each slot from the list (using the key id and the pin). In most cases we will see one slot and try once.